src/runtime/program/vote/execute.zig

```diff
@@ -138,10 +174,287 @@ fn intializeAccount(
+///
+/// Authorize the given pubkey to withdraw or sign votes. This may be called multiple times,
+/// but will implicitly withdraw authorization from the previously authorized
+/// key.
+fn authorize(
+    allocator: std.mem.Allocator,
+    ic: *InstructionContext,
+    vote_account: *BorrowedAccount,
+    authorized: Pubkey,
+    vote_authorize: VoteAuthorize,
+    clock: Clock,
+    signers: ?std.AutoHashMap(Pubkey, void),
+) InstructionError!void {
+    const versioned_state = try vote_account.deserializeFromAccountData(
+        allocator,
+        VoteStateVersions,
+    );
+    var vote_state = versioned_state.convertToCurrent(allocator) catch {
+        // TODO okay to convert out of memory to InvalidAccountData?
+        return InstructionError.InvalidAccountData;
+    };
+    defer vote_state.deinit();
+
+    switch (vote_authorize) {
+        .voter => {
+            const current_epoch = clock.epoch;
+            const target_epoch = std.math.add(u64, current_epoch, 1) catch {
+                return InstructionError.InvalidAccountData;
+            };
+
+            try checkAuthority(
+                allocator,
+                ic,
+                &vote_state,
+                signers,
+                current_epoch,
+            );
+
+            vote_state.setNewAuthorizedVoter(
+                authorized,
+                target_epoch,
+            ) catch |err| {
+                switch (err) {
+                    VoteError.TooSoonToReauthorize => {
+                        ic.tc.custom_error = @intFromError(err);
+                        return InstructionError.Custom;
+                    },
+                    else => {
+                        return InstructionError.InvalidAccountData;
+                    },
+                }
+            };
+        },
+        .withdrawer => {
+            // current authorized withdrawer must say "yay".
+            const authorized_withdrawer_signer = if (signers) |signers_|
+                try verifyAuthorizedSigner(vote_state.authorized_withdrawer, signers_)
+            else
+                ic.info.isPubkeySigner(vote_state.authorized_withdrawer);
+
+            if (!authorized_withdrawer_signer) {
+                return InstructionError.MissingRequiredSignature;
+            }
+            vote_state.authorized_withdrawer = authorized;
+        },
+    }
+    try vote_account.serializeIntoAccountData(VoteStateVersions{ .current = vote_state });
+}
+
+/// Analogous to
+/// https://github.com/anza-xyz/agave/blob/49fb51295c1062b6b09e585b2fe0a4676c33d3d4/programs/vote/src/vote_state/mod.rs#L691-L692
+/// https://github.com/anza-xyz/agave/blob/49fb51295c1062b6b09e585b2fe0a4676c33d3d4/programs/vote/src/vote_state/mod.rs#L701-L707
+/// https://github.com/anza-xyz/solana-sdk/blob/4e30766b8d327f0191df6490e48d9ef521956495/vote-interface/src/state/mod.rs#L873
+///
+/// Check if the current authorized withdrawer or epoch authorized voter is a signer.
+fn checkAuthority(
```

Comment:
```suggestion
fn validateAuthorityIsSigner(
```

i think `validate` is the word we want to use going forward - we talked about this in the past in a random PR i think - @yewman is this correct?

Reply:
Actually had it with `validate` prefix but changed it when I noticed this https://github.com/Syndica/sig/blob/3516baf98b64bd8c2a6f63e14703927ddf8244c5/src/runtime/program/system_program/execute.zig#L782 in the system program

Reply:
huh yeah that ref should be validate too imo